c++/src/kj/compat/http.c++

```diff
@@ -6307,31 +6462,69 @@ public:
+                            kj::AsyncIoStream& connection,
+                            ConnectResponse& response) override {
     KJ_REQUIRE(!headers.isWebSocket(), "WebSocket upgrade headers are not permitted in a connect.");
-    return client.connect(host, headers)
-        .then([&tunnel](HttpClient::ConnectResponse&& response) -> kj::Promise<void> {
-      KJ_SWITCH_ONEOF(response.connectionOrBody) {
-        KJ_CASE_ONEOF(io, kj::Own<kj::AsyncIoStream>) {
-          auto io2 = tunnel.accept(response.statusCode, response.statusText, *response.headers);
-          auto promises = kj::heapArrayBuilder<kj::Promise<void>>(2);
-
-          promises.add(io->pumpTo(*io2).then([&io2 = *io2](uint64_t size) {
-            io2.shutdownWrite();
-          }).eagerlyEvaluate(nullptr));
-          promises.add(io2->pumpTo(*io).then([&io = *io](uint64_t size) {
-            io.shutdownWrite();
-          }).eagerlyEvaluate(nullptr));
-          return kj::joinPromises(promises.finish()).attach(kj::mv(io), kj::mv(io2));
-        }
-        KJ_CASE_ONEOF(input, kj::Own<kj::AsyncInputStream>) {
-          auto io2 = tunnel.reject(response.statusCode, response.statusText, *response.headers);
-          return input->pumpTo(*io2).ignoreResult().attach(kj::mv(input), kj::mv(io2));
+
+    auto request = client.connect(host, headers);
+
+    // This operates optimistically. In order to support pipelining, we connect the
+    // input and outputs streams immediately, even if we're not yet certain that the
+    // tunnel can actually be established.
+    auto promises = kj::heapArrayBuilder<kj::Promise<void>>(3);
+
+    // For the inbound pipe (from the clients stream to the passed in stream)
+    // We want to guard reads pending the acceptance of the tunnel. If the
+    // tunnel is not accepted, the guard will be rejected, causing pending
+    // reads to fail.
+    auto paf = kj::newPromiseAndFulfiller<kj::Maybe<HttpInputStreamImpl::ReleasedBuffer>>();
+    auto io = kj::heap<AsyncIoStreamWithGuards>(
+        kj::mv(request.connection),
+        kj::mv(paf.promise) /* read guard */,
+        kj::READY_NOW /* write guard */);
+
+    // Writing from connection to io is unguarded and allowed immediately.
+    promises.add(connection.pumpTo(*io).then([&io=*io](uint64_t size) {
+      io.shutdownWrite();
+    }).eagerlyEvaluate(nullptr));
+
+    promises.add(io->pumpTo(connection).then([&connection](uint64_t size) {
+      connection.shutdownWrite();
+    }).eagerlyEvaluate(nullptr));
+
+    promises.add(request.status.then(
+        [&response,&io=*io,&connection,fulfiller=kj::mv(paf.fulfiller)]
+        (HttpClient::ConnectRequest::Status status) mutable -> kj::Promise<void> {
+      if (status.statusCode >= 200 && status.statusCode < 300) {
+        // Release the read guard!
+        fulfiller->fulfill(kj::Maybe<HttpInputStreamImpl::ReleasedBuffer>(nullptr));
+        response.accept(status.statusCode, status.statusText, *status.headers);
```

Comment:
Do we need a `return kj::READY_NOW` statement here to avoid the `KJ_UNREACHABLE` below?

Reply:
I've added it in https://github.com/capnproto/capnproto/pull/1586